dingo/gw/noise/synthetic_noise/asd_sampling.py

```diff
@@ -0,0 +1,112 @@
+            x_positions = param_dict["x_positions"]
+            spectral_features = param_dict["spectral_features"]
+
+            for i in range(spectral_features.shape[1]):
+                try:
+                    spectral_kde = stats.gaussian_kde(
+                        spectral_features[:, i, :].T,
+                        bw_method=float(self.settings["bandwidth_spectral"]),
+                        weights=weights,
+                    )
+                except np.linalg.LinAlgError:
+                    print(
+                        "Warning: Singular Matrix encountered in spectral KDE. Adding small Gaussian noise..."
+                    )
+                    perturbed_features = spectral_features[:, i, :] + np.random.normal(
+                        0, 1.0e-4, size=spectral_features[:, i, :].shape
+                    )
+                    spectral_kde = stats.gaussian_kde(
+                        perturbed_features.T,
+                        bw_method=float(self.settings["bandwidth_spectral"]),
+                        weights=weights,
+                    )
+                self.spectral_kde[det].append(spectral_kde)
+
+            split_indices = [0, len(x_positions)]
+            split_indices += [
+                get_index_for_elem(x_positions, f) + 1
+                for f in self.settings["split_frequencies"]
+            ]
+            split_indices = sorted(split_indices)
+
+            for i in range(len(split_indices) - 1):
+                vals = y_values[:, split_indices[i]: split_indices[i + 1]].T
+                kde_vals = stats.gaussian_kde(
+                    vals, bw_method=float(self.settings["bandwidth_spline"])
+                )
+                self.broadband_kde[det].append(kde_vals)
+
+    def sample(self, rescaling_ys):
+
+        num_samples = self.settings["num_samples"]
+
+        parameters_dicts = copy.deepcopy(self.parameter_dicts)
+        asds_dict = {}
+
+        for det, param_dict in self.parameter_dicts.items():
+            asds_dict[det] = []
+
+            features = np.zeros((num_samples, len(self.spectral_kde[det]), 3))
+            for i, kde in enumerate(self.spectral_kde[det]):
+                features[:, i, :] = kde.resample(size=num_samples).T
+
+            parameters_dicts[det]["spectral_features"] = features
+
+            y_values = np.zeros((num_samples, 0))
+
+            for i, kde in enumerate(self.broadband_kde[det]):
+                y_values = np.concatenate(
+                    (y_values, kde.resample(size=num_samples).T), axis=1
+                )
+            # rescale base noise
+            if rescaling_ys:
+                y_values_mean = np.mean(y_values, axis=0)
+                y_values = (
+                    y_values - y_values_mean[None, :] + rescaling_ys[det]
+                )
+            parameters_dicts[det]["y_values"] = y_values
+
+        return parameters_dicts
+
+
+def get_rescaling_params(filenames, parameterization_settings):
```

Comment:
A docstring would be good here. It's not clear, e.g., what the `filenames` should be (i.e., ASDDataset files).